runtime/vm/jvminit.c

```diff
@@ -1868,7 +1868,35 @@ IDATA VMInitStages(J9JavaVM *vm, IDATA stage, void* reserved) {
 				}
 			}
 #endif /* !defined(WIN32) && !defined(J9ZTPF) */
+			/* Parse options related to Large Pages */
+			{
+				IDATA argIndexUseLargePages = FIND_AND_CONSUME_ARG(EXACT_MATCH, MAPOPT_XXUSELARGEPAGES, NULL);
+				IDATA argIndexLargePageSizeInBytes = FIND_AND_CONSUME_ARG(STARTSWITH_MATCH, MAPOPT_XXLARGEPAGESIZEINBYTES_EQUALS, NULL);
```

Comment:
The approach taken here is to decide which of the two new arguments has priority and only record it.  It leaves the priority of this option versus the old `-Xlp`-style options to be re-implemented by the consumers.

I would prefer to have that decision centralized here rather than spread out and potentially handled inconsistently in the different components due to future changes.

We have an some macros that make it easy to find an option and then to iteratively search forward by the different matches.  See an example in the verifier option handling: [1]

The idea would be to do something similar and find the first `-Xlp*` option here using a `STARTSWITH_MATCH` and then marking each `-Xlp` option that occurs before the new options as "consumed" using the `CONSUME_ARG(j9vm_args, element)` macro.

JIT and GC processing then would only have 1 set of data to examine - either the new options data or the old `Xlp` style data and every component would have a consistent view of the data.

[1] https://github.com/eclipse/openj9/blob/d6506efbb6f74e7826f7c825c76489ab6e8ba944/runtime/bcverify/bcverify.c#L2694-L2713

Reply:
This is certainly a great idea. There's just one caveat.

Since these -XX options impact both the codecache and objectheap then -Xlp options can be ignored if they precede these options.
The opposite is not true, since -Xlp, -Xlp:codecache, and -Xlp:objectheap selectively impact either the objectheap, or codecache only. It is assumed that the respective components handle these partial options.

The changes I've made allow for the codecache and objectheap to handle any combination of `-Xlp*, XX:+UseLargePages, and -XX:LargePageInBytes=<size>`.
So this could still work, given that if a combination of -Xlp, and -XX options are passed and if the right most option is -Xlp then we can't simply consume the -XX page options unless we start keeping track of for which -Xlp option impacts the codecache (-Xlp<size> -Xlp:codecache), and which impact (-Xlp, -Xlp<size>, -Xlp:objectheap). 

An example of this is using the following:
`-XX:+UseLargePages -Xlp:objectheap:pagesize=2G`
Codecache would use 1M pages, while objectheap would use 2G pages.

Should we disallow the user from using a combination of -XX page options and -Xlp options entirely?

Reply:
I suspect the best approach is to keep track of options that impact the `objectheap`, and `codecache` separately. Then consume all options that don't apply. Meaning the worst case scenario would be that two options would not be consumed.

Reply:
Code for my suggestion. 

```
/* Parse options related to Large Pages */
			{
				IDATA codecacheIndex = -1;
				IDATA objectheapIndex = -1;
				IDATA index = j9vm_args->nOptions;
				char *option = NULL;

				while (index != -1) {

					IDATA argIndexUseLargePages = FIND_NEXT_ARG_IN_VMARGS(EXACT_MATCH, MAPOPT_XXUSELARGEPAGES, NULL, index);
					IDATA argIndexDontUseLargePages = FIND_NEXT_ARG_IN_VMARGS(EXACT_MATCH, MAPOPT_XXDISABLEUSELARGEPAGES, NULL, index);
					IDATA argIndexLargePageSizeInBytes = FIND_NEXT_ARG_IN_VMARGS(EXACT_MATCH, MAPOPT_XXLARGEPAGESIZEINBYTES_EQUALS, NULL, index);
					IDATA argIndexXlp = FIND_NEXT_ARG_IN_VMARGS(STARTSWITH_MATCH, "-Xlp", NULL, index);
					BOOLEAN isConsumable = FALSE;

					/* Find Maximum index */
					index = MAX(argIndexUseLargePages, argIndexDontUseLargePages);
					index = MAX(index, argIndexLargePageSizeInBytes);
					index = MAX(index, argIndexXlp);

					/* No large page option found. Exit */
					if (-1 == index) {
						break;
					}

					option = getOptionString(j9vm_args, index);

					/* -XX:[+/-]UseLargePages & -XX:LargePageSizeInBytes=<size> & Xlp<size> */
					if (((0 == strcmp(option, MAPOPT_XXUSELARGEPAGES)) || (0 == strcmp(option, MAPOPT_XXDISABLEUSELARGEPAGES))) ||
						(NULL != strstr(option, MAPOPT_XXLARGEPAGESIZEINBYTES_EQUALS)) ||
						((NULL != strstr(option, "-Xlp")) && (0 > strmcp(option, "-Xlp")))) {
						if ((-1 == codecacheIndex) || (-1 == objectheapIndex)) {
							codecacheIndex = codecacheIndex == -1? index: codecacheIndex;
							objectheapIndex = objectheapIndex == -1? index: objectheapIndex;
						}
						else {
							isConsumable = TRUE;
						}
					}
					/* -Xlp:codecache */
					else if (NULL != strstr(option, "-Xlp:codecache")) {
						if (-1 == codecacheIndex) {
							codecacheIndex = index;
						}
						else {
							isConsumable = TRUE;
						}
					}
					/* -Xlp:objectheap & -Xlp */
					else if (NULL != strstr(option, "Xlp:objectheap") || (0 == strmcp(option, "-Xlp"))) {
						if (-1 == objectheap) {
							objectheapIndex = index;
						}
						else {
							isConsumable = TRUE;
						}
					}

					/* Consume argument */
					if (isConsumable) {
						CONSUME_ARG(vm->vmArgsArray, index);
					}
				}
			}
```

Reply:
@DanHeidinga 

>Since these -XX options impact both the codecache and objectheap then -Xlp options can be ignored if they precede these options.
The opposite is not true, since -Xlp, -Xlp:codecache, and -Xlp:objectheap selectively impact either the objectheap, or codecache only. It is assumed that the respective components handle these partial options.

As I've mentioned. It's not possible to isolate between either using -Xlp:* options or the -XX options unless we make a decision to ignore all -Xlp:* options when -XX options are present no matter of the index within vmargs.

The closest thing we can do is flag to ignore options that will not be used within the codecache or the objectheap. In this case all -Xlp options will be ignored, if any of the -XX options are used at a later index. 
Examples
 -Xlp<size> -XX:+UseLargePages.  (-Xlp<size> is ignored)
-Xlp:+UseLargePages -Xlp:codecache:pagesize=<size> (No option is ignored. UseLargePages will be used by the objectheap, and -Xlp:codecache used by the codecache).